Userland/cal.cpp

```diff
@@ -21,29 +36,139 @@ int get_number_of_days(int month, int year)
     return is_long_month ? 31 : 30;
 }
 
+void append_to_print(char* buffer, int row, int column, char* text)
+{
+    int starting_point = (line_width * row) + (column * column_width);
+    for (int i = 0; text[i] != '\0'; i++) {
+        buffer[starting_point + i] = text[i];
+    }
+}
+
+void insert_month_to_print(int column, int month, int year)
+{
+    int printing_column = column;
+    int printing_row = 0;
+
+    // FIXME: Both the month name and month header text should be provided by a locale
+    sprintf(temp_buffer, "     %02u - %04u    ", month, year);
+    append_to_print(print_buffer, printing_row, printing_column, temp_buffer);
+    printing_row++;
+
+    sprintf(temp_buffer, "Su Mo Tu We Th Fr Sa");
+    append_to_print(print_buffer, printing_row, printing_column, temp_buffer);
+    printing_row++;
+
+    int day_to_print = 1;
+    int first_day_of_week_for_month = day_of_week(1, month, year);
+    int days_in_the_month = get_number_of_days(month, year);
+    int last_written_chars = 0;
+    for (int i = 1; day_to_print <= days_in_the_month; ++i) {
+        if (i - 1 < first_day_of_week_for_month) {
+            last_written_chars += sprintf(temp_buffer + last_written_chars, "   ");
+        } else {
+            if (year == current_year && month == current_month && target_day == day_to_print) {
+                // FIXME: To replicate Unix cal it would be better to use "\x1b[30;47m%2d\x1b[0m " in here instead of *.
+                //        However, doing that messes up the layout.
```

Comment:
You mean the layout inside `print_buffer`, not that the terminal would display this incorrectly? Then maybe the print buffer should be stored differently, perhaps as a `Vector<` of `String>`s?..

Reply:
What if you structure it like this?

```cpp
class Month {
    Vector<String> render() const;
};

class Year {
    Vector<String> render() const;
};
```

`Year::render()` would call `render()` on its months and then join the lines they produce.